nixos/modules/programs/tsm-client.nix

```diff
@@ -105,6 +106,19 @@ let
     };
     config.commmethod = mkDefault "v6tcpip";  # uses v4 or v6, based on dns lookup result
     config.passwordaccess = if config.genPasswd then "generate" else "prompt";
+    # XXX migration code for freeform settings, these can be removed after 23.11:
+    options.warnings = optionsGlobal.warnings;
+    options.assertions = optionsGlobal.assertions;
```

Comment:
Such an issue exists already: https://github.com/NixOS/nixpkgs/issues/96006

I was thinking about a general solution, but the tricky part is "transporting" assertions and warning out of the submodule into the main configuration area.  Here, in the PR at hand, I'm doing it manually, by defining `assertions` and `warnings` options in the submodule.  Then I have to evaluate both of them outside of the submodule (with `mapAttrsToList`...) and I have to filter them out of the submodule's result where it is used to generate the configuration.  I ultimatelly failed at encapsulating all this inside a `mkRenamedAttrsOptionModule`, it just requires too many assumptions about the design and usage of the submodule where it would be applied.

Note: I also encountered these PRs that try/tried to implement `assertions` and `warnings` at submodule level.  Such an addition would probably be a good step towards a general solution for submodules.

*   https://github.com/NixOS/nixpkgs/pull/97023 , got reverted due to infinite recursion errors, has already many links to other PRs where it would be useful.
*   https://github.com/NixOS/nixpkgs/pull/207187